app/services/subscriptions/dates/quarterly_service.rb

```diff
@@ -0,0 +1,143 @@
+      end
+
+      def compute_charges_from_date
+        if terminated?
+          return subscription.anniversary? ? previous_anniversary_day(billing_date) : billing_date.beginning_of_quarter
+        end
+
+        return compute_from_date if plan.pay_in_arrear?
+        return base_date.beginning_of_quarter if calendar?
+
+        previous_anniversary_day(base_date)
+      end
+
+      def compute_charges_to_date
+        return compute_charges_from_date.end_of_quarter if calendar?
+
+        compute_to_date(compute_charges_from_date)
+      end
+
+      def compute_duration(from_date:)
+        next_to_date = compute_to_date(from_date)
+
+        (next_to_date.to_date + 1.day - from_date.to_date).to_i
+      end
+
+      alias compute_charges_duration compute_duration
+
+      def compute_base_date
+        billing_date - 3.months
+      end
+
+      def compute_to_date(from_date = compute_from_date)
+        if subscription.calendar? || (subscription_at.day == 1 && [1, 4, 7, 10].include?(subscription_at.month))
+          return from_date.end_of_quarter
+        end
+
+        year = from_date.year
+        month = from_date.month + 3
+        day = subscription_at.day - 1
+
+        if month > 12
+          month = (month % 12) == 0 ? 12 : (month % 12)
+          year += 1
+        end
+
+        build_date(year, month, day)
+      end
+
+      def compute_next_end_of_period
+        return billing_date.end_of_quarter if calendar?
+
+        year = billing_date.year
+        month = billing_date.month
+        day = subscription_at.day
+
+        # NOTE: we need the last day of the period, and not the first of the next one
+        result_date = build_date(year, month, day) - 1.day
+        return result_date if result_date >= billing_date
+
+        month += 3
+        if month > 12
+          month = (month % 12) == 0 ? 12 : (month % 12)
+          year += 1
+        end
+
+        build_date(year, month, day) - 1.day
+      end
+
+      def compute_previous_beginning_of_period(date)
+        return date.beginning_of_quarter if calendar?
+
+        previous_anniversary_day(date)
+      end
+
+      def previous_anniversary_day(date)
+        year = nil
+        month = nil
+
+        # NOTE: in anniversary mode if both subscription date and current date are
+        #       the last day of month, anniversary day is on the current day
+        day = if subscription.anniversary? && last_day_of_month?(date) && last_day_of_month?(subscription_at)
+          date.day
+        else
+          subscription_at.day
+        end
+
+        billing_months = [
+          (subscription_at.month % 12) == 0 ? 12 : (subscription_at.month % 12),
+          ((subscription_at.month + 3) % 12) == 0 ? 12 : ((subscription_at.month + 3) % 12),
+          ((subscription_at.month + 6) % 12) == 0 ? 12 : ((subscription_at.month + 6) % 12),
+          ((subscription_at.month + 9) % 12) == 0 ? 12 : ((subscription_at.month + 9) % 12),
```

Comment:
Style/NumericPredicate: Use `((subscription_at.month + 9) % 12).zero?` instead of `((subscription_at.month + 9) % 12) == 0`. (https://rubystyle.guide#predicate-methods)